### ISSUE TITLE
S3Client.headObject does not throw NoSuchKeyException as documented

### ISSUE THREAD

**Issue by ghadishayban** [aws/aws-sdk-java-v2#544]
## Expected Behavior
The docs say that headObject should return a NoSuchKeyException when not found.

## Current Behavior
It does not throw NSKE. Instead, it throws an ordinary:
```
S3Exception null (Service: S3Client; Status Code: 404; Request ID: null)  software.amazon.awssdk.core.http.pipeline.stages.HandleResponseStage.handleErrorRes\
ponse (HandleResponseStage.java:114)
```

https://github.com/aws/aws-sdk-java-v2/issues/147 appeared to fix this for getObject, but headObject still busted (there's even a test for it in that changeset.) 

## Steps to Reproduce (for bugs)
HEAD a non existent key with the S3Client

## Your Environment
JDK 10 + AWS SDK v2 preview 10

**Comment by ghadishayban** [comment]
I will add, I really don't care for throwing exceptions in such a common case. HEAD will either return stat meta or not.

**Comment by codylerum** [comment]
Also see #297

**Comment by zoewangg** [comment]
This has been fixed as part of #830 and will be included in the next release.
